drivers/gpio/gpio_sx1509b.c

```diff
@@ -112,6 +148,83 @@ static inline int i2c_reg_write_word_be(struct device *dev, u16_t dev_addr,
 	return i2c_write(dev, tx_buf, 3, dev_addr);
 }
 
+/**
+ * @brief Write a big-endian byte to an internal address of an I2C slave.
+ *
+ * @param dev Pointer to the device structure for the driver instance.
+ * @param dev_addr Address of the I2C device for writing.
+ * @param reg_addr Address of the internal register being written.
+ * @param value Value to be written to internal register.
+ *
+ * @retval 0 If successful.
+ * @retval -EIO General input / output error.
+ */
+static inline int i2c_reg_write_byte_be(struct device *dev, u16_t dev_addr,
+					u8_t reg_addr, u8_t value)
+{
+	u8_t tx_buf[3] = { reg_addr, value };
+
+	return i2c_write(dev, tx_buf, 2, dev_addr);
+}
+
+#ifdef CONFIG_GPIO_SX1509B_INTERRUPT
+static int sx1509b_handle_interrupt(void *arg)
+{
+	struct device *dev = (struct device *) arg;
+	const struct sx1509b_config *cfg = dev->config->config_info;
+	struct sx1509b_drv_data *drv_data = dev->driver_data;
+	int ret = 0;
+	u16_t int_source;
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	ret = i2c_burst_read(drv_data->i2c_master, cfg->i2c_slave_addr,
+				 SX1509B_REG_INTERRUPT_SOURCE,
+				 (u8_t *)&int_source,
+				 sizeof(int_source));
+	if (ret) {
+		goto out;
+	}
+
+	int_source = sys_be16_to_cpu(int_source);
+
+	if ((int_source & drv_data->cb_pins) != 0) {
```

Comment:
You can't invoke the callbacks with the semaphore held, because the callbacks might (generally will) use the GPIO device to reconfigure the interrupt conditions.